src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets

```diff
@@ -243,19 +257,39 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         <Message Text="CppWinRTMdMergeMetadataDirectories: @(CppWinRTMdMergeMetadataDirectories)" Importance="$(CppWinRTVerbosity)"/>
     </Target>
 
-    <!--If building any Xaml content, include XamlMetaDataProvider in project metadata-->
-    <Target Name="CppWinRTAddXamlMetaDataProviderIdl" Condition="'$(XamlLanguage)' == 'CppWinRT' and '$(CppWinRTAddXamlMetaDataProviderIdl)' == 'true'">
+    <!-- Adds the XamlMetadataProvider idl to the Midl itemgroup, if building any xaml content -->
+    <Target Name="CppWinRTComputeXamlGeneratedMidlInputs"
+            DependsOnTargets="$(CppWinRTComputeXamlGeneratedMidlInputsDependsOn)"
+            Condition="'@(Page)@(ApplicationDefinition)' != '' and '$(XamlLanguage)' == 'CppWinRT' and '$(CppWinRTAddXamlMetaDataProviderIdl)' == 'true'">
         <PropertyGroup>
-            <XamlMetaDataProviderIdl>
-                $(IntDir)XamlMetaDataProvider.idl
-            </XamlMetaDataProviderIdl>
-            <XamlMetaDataProviderCpp>
-                $(GeneratedFilesDir)XamlMetaDataProvider.cpp
-            </XamlMetaDataProviderCpp>
-            <_HasXaml>false</_HasXaml>
-            <_HasXaml Condition="'@(Page)' != '' Or '@(ApplicationDefinition)' != ''">true</_HasXaml>
+            <_DisableReferences>false</_DisableReferences>
+            <_DisableReferences Condition="('$(CppWinRTOverrideSDKReferences)' != 'true') and ('$(TargetPlatformVersion)' &lt; '10.0.18310.0')">true</_DisableReferences>
         </PropertyGroup>
-        <PropertyGroup Condition="$(_HasXaml)">
+        
+        <ItemGroup>
+            <Midl Remove="$(XamlMetaDataProviderIdl)" />
+            <Midl Include="$(XamlMetaDataProviderIdl)">
+                <DisableReferences Condition="$(_DisableReferences)">>true</DisableReferences>
+            </Midl>
+        </ItemGroup>
+    </Target>
+
+    <!-- Adds the XamlMetadataProvider cpp to the ClCompile itemgroup, if building any xaml content -->
+    <Target Name="CppWinRTComputeXamlGeneratedCompileInputs"
+            DependsOnTargets="$(CppWinRTComputeXamlGeneratedCompileInputsDependsOn)"
+            Condition="'@(Page)@(ApplicationDefinition)' != '' and '$(XamlLanguage)' == 'CppWinRT' and '$(CppWinRTAddXamlMetaDataProviderIdl)' == 'true'">
+        <ItemGroup>
+            <ClCompile Remove="$(XamlMetaDataProviderCpp)" />
+            <ClCompile Include="$(XamlMetaDataProviderCpp)" Condition="'$(CppWinRTOptimized)'=='true'">
+                <CompilerIteration>XamlGenerated</CompilerIteration>
+            </ClCompile>
+        </ItemGroup>
+    </Target>
+
+    <!--If building any Xaml content, write XamlMetaDataProvider idl file -->
+    <Target Name="CppWinRTAddXamlMetaDataProviderIdl"
+            Condition="'@(Page)@(ApplicationDefinition)' != '' and '$(XamlLanguage)' == 'CppWinRT' and '$(CppWinRTAddXamlMetaDataProviderIdl)' == 'true'">
+        <PropertyGroup>
             <_DisableReferences>false</_DisableReferences>
```

Comment:
still need to clean up _DisableReferences here

Reply:
What exactly needs to be cleaned up here?

Reply:
ah, never mind - I didn't see the use of it in XamlMarkupIdlImport

---
In reply to: [330668339](https://github.com/microsoft/xlang/pull/609#discussion_r330668339) [](ancestors = 330668339)